### PR TITLE
moe: integrate EXL3 Trellis into the fused W4A16 API

### PR DESCRIPTION
**Pull request by voipmonitor**
## Summary

Integrate EXL3 Trellis/MCG as a W4A16 weight format under the existing planned `sparkinfer.moe.fused_moe` API.

This supersedes #49's separate public `sparkinfer.moe.trellis_moe` surface while retaining its Trellis intrinsics, native weight preparation, full-rotation kernel path, graph-safe route packing, and dense kernel support.

## Why this shape

`trellis3_t256`, `nf3_2p1`, packed FP4, and ModelOpt NVFP4 all execute through the same fused-TP W4A16 strategy. The public API should therefore vary the weight recipe, not create one MoE operator per storage format.

The integration now uses one lifecycle:

- `plan_weights(source_format="exl3_trellis_mcg", quant_modes="w4a16", ...)`
- `prepare_weights(...)`
- `Caps(..., quant_mode="w4a16")`
- `plan(...)`
- caller-owned `scratch_specs()` / `bind(...)`
- `run(...)`

Trellis and NF3 share `PreparedW4A16MoeWeights`, `TPMoEScratchPlan`, the caller-owned scratch arena, cache clearing, and activation dispatch. No second arena or parallel bind/run implementation remains.

## Functional changes

- Add Trellis 3/4/5/6-bpw FP16/BF16 decode intrinsics.
- Add native Trellis256 MoE and dense weight preparation.
- Add `trellis3_t256` full-rotation W4A16 execution with FP32 route accumulation.
- Support both SiLU and SiTU through the standard fused-MoE activation contract.
- Keep persistent prepared weights free of duplicate scratch storage.
- Add public dense API at `sparkinfer.gemm.trellis_linear`; keep `_shared.kernels.w4a16` private.
- Gate support through the existing SM120/SM121 and CUTLASS DSL capability checks.
- Document why the kernel workspace is zeroed at bind time.
- Preserve the scheduler-planned route geometry through scratch binding and execution. The live batch heuristic may choose a different `block_m`; it must not silently invalidate the planned route and expert-block capacities.

## Review decomposition

Independent changes from #49 were split out:

- #88: compiler device/architecture compile-cache identity
- #89: high-i64 page-offset indexer regression test

This PR contains only the Trellis compute and unified MoE/dense API work.

## Validation

On the pinned CUDA 13.2 / SM120 v20-r11 runtime:

- 122 passed: Trellis fused-MoE/dense correctness, SiLU/SiTU, BF16/FP16, small-M partial blocks, below-capacity CUDA graph replay, shared execution/scratch lifecycle, route packing/capacity diagnostics, and public registry/API contracts.
- 19 passed in the paired vLLM #190 port tests.
- Unified runtime import smoke test confirms vLLM resolves `sparkinfer.moe.fused_moe`; no `trellis_moe` compatibility surface is required.
- A real layer-3 rank-0 checkpoint test used native Trellis tensors at production `H=6144`, `I=512`, executed the unified plan with zero-copy weights, and passed eager plus CUDA graph replay. Independent unified runs were bitwise identical. Against the r11 legacy `trellis_moe` runtime on the same deterministic input, relative L2 error was `2.01e-4`, cosine similarity `0.999999881`, and maximum absolute difference `5.65e-5` (the comparison spans intervening shared-W4A16 changes on `master`, not only the API wrapper).
- The full 295 GiB `brandonmusic/GLM-5.2-EXL3-TR3-3.0bpw` checkpoint passed TP4/DCP4/MTP3 load, memory profiling, piecewise/full CUDA graph capture, MTP capture, and a real HTTP generation (`EXL3_OK`). At `GPU_MEMORY_UTILIZATION=0.90`, the run reported 735,232 KV-cache tokens and a 414.7 MiB planned EXL3 prefill arena per rank.
- This E2E run exposed and now covers the geometry contract: planning selected `block_m=64`, while the live `m=1024`, top-k 8, 256-expert heuristic preferred 48. The binding now carries the planned value into route packing and GEMM; disagreement with a supplied launch fails before execution. A direct GPU check also verified that 48 and 64 compile to distinct kernels, ruling out a compile-cache collision.
- Ruff and `git diff --check` pass for every changed file.

## Consumer migration

The paired vLLM #190 update changes its EXL3 backend from `trellis_moe` to the unified weight-plan and fused-MoE API. `trellis_moe` was unreleased, so no deprecated compatibility shim is added.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by brandonmmusic-max** [comment]
Luke made a long post of changes to be made on #49 .   Should I port that suggested work to this PR?  Let me know or if you’d already done it!

File: sparkinfer/moe/_shared/kernels/w4a16/kernel.py [hunk 1]
```diff
@@ -1381,6 +1531,8 @@ def _read_moe_block_data(
                 ].to(Int32)
                 st_shared_i32(smem_base + Int32(self.sh_route_off * 16), idx)
                 rd_row = idx // Int32(self.top_k)
+                if cutlass.const_expr(self.route_major_a):
+                    rd_row = idx
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/_shared/kernels/w4a16/kernel.py:1535]: Fixed in 987b829. Route/row IDs and row strides are widened before every affected A/C element-offset multiplication, and the route-major CuTe tensor extents are Int64. The actual production-shaped EXL3 layer compiled and ran on SM120 after the change.

File: sparkinfer/moe/_shared/kernels/w4a16/kernel.py [hunk 2]
```diff
@@ -4290,7 +4661,49 @@ def __init__(
...
+        if self.intermediate_rotation:
+            if weight_layout != "trellis3_t256":
+                raise ValueError(
+                    "intermediate_rotation is only supported for trellis3_t256"
+                )
+            if not is_gated or self.activation_is_swigluoai or self.has_swiglu_limit:
+                raise ValueError(
+                    "intermediate_rotation requires plain gated silu "
+                    "(no swiglu limit/oai)"
+                )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/_shared/kernels/w4a16/kernel.py:4681]: Fixed in 987b829. The diagnostic now accurately names unclamped gated SiLU or SiTU and continues to reject the SwigluOAI/limited variants.

File: sparkinfer/moe/_shared/kernels/w4a16/kernel.py [hunk 3]
```diff
@@ -7233,6 +8404,20 @@ def _w4a16_small_m_direct_launch_fake(
...
+
+def _rot_scales_dummy(device: torch.device) -> torch.Tensor:
+    """Tiny fp16 placeholder for the rot_scales kernel slot when the
+    intermediate-rotation epilogue is disabled (never dereferenced by the
+    const_expr-gated kernel; keeps one compiled signature across all layouts)."""
+    t = _ROT_SCALES_DUMMY.get(device)
+    if t is None:
+        t = torch.zeros(1, dtype=torch.float16, device=device)
+        _ROT_SCALES_DUMMY[device] = t
+    return t
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/_shared/kernels/w4a16/kernel.py:8431]: Fixed in 987b829. The per-device placeholder is materialized by planned-launch prewarm; a first lookup during capture now fails clearly instead of allocating. A regression test covers the capture guard.

File: sparkinfer/moe/_shared/kernels/w4a16/kernel.py [hunk 4]
```diff
@@ -8000,9 +9336,259 @@ def pack_topk_routes_by_expert(
...
+
+    hadamard_128 = _resolve_exl3_hadamard_128(hadamard_128)
+    x_f16 = x if x.dtype == torch.float16 else x.to(torch.float16)
+    rotated_f16 = torch.empty_like(x_f16)
+    hadamard_128(x_f16, rotated_f16, prepared_dense.suh, None, 1.0)
+    rotated_compute = (
+        rotated_f16
+        if compute_dtype == torch.float16
+        else rotated_f16.to(torch.bfloat16)
+    )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/_shared/kernels/w4a16/kernel.py:9518]: Fixed in 987b829. The dense public API now accepts caller-owned input-conversion, rotation, compute-cast, GEMM-cast, and output-rotation buffers. Fallback allocation is eager-only; capture fails closed if required storage is absent. A real SM120 BF16 CUDA-graph replay test reuses all buffers.

File: sparkinfer/moe/fused_moe/_impl.py [hunk 5]
```diff
@@ -2237,6 +2279,50 @@ def _build_tp_moe_fp4_binding_from_views(
...
+                continue
+            if (
+                expert_map.dtype != torch.int32
+                or expert_map.device != a.device
+                or tuple(expert_map.shape) != (plan.route_E,)
+                or not expert_map.is_contiguous()
+            ):
+                raise ValueError(
+                    f"{name} must be contiguous int32[{plan.route_E}] on {a.device}"
+                )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/fused_moe/_impl.py:2316]: Fixed in 987b829. Non-full-rotation plans now reject either expert-map argument instead of silently discarding it, with a regression test.

File: sparkinfer/moe/fused_moe/_impl.py [hunk 6]
```diff
@@ -4666,6 +4836,119 @@ def prepare_sparkinfer_fp4_moe_weights(
...
+        if trellis_mcg is not None:
+            marker = (
+                int(trellis_mcg.item())
+                if isinstance(trellis_mcg, torch.Tensor)
+                else int(trellis_mcg)
+            ) & 0xFFFFFFFF
+            if marker != 0xCBAC1FED:
+                raise ValueError(
+                    f"unexpected EXL3 MCG marker {marker:#010x}; expected 0xcbac1fed"
+                )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/fused_moe/_impl.py]: Fixed in 987b829. EXL3 MCG preparation now requires the checkpoint marker and validates the exact MCG value before selecting the MCG decoder. Missing and invalid marker tests are included.

File: sparkinfer/moe/fused_moe/_impl.py [hunk 7]
```diff
@@ -6536,12 +6873,54 @@ def build_tp_moe_fp4_binding(
         elif quant_mode == "w4a16":
             weight_layout = _w4a16_weight_layout_for_source(source_format)
             scale_format = _w4a16_scale_format_for_source(source_format)
+        if workspace.full_rotation:
+            if workspace.full_rotation_output is None:
+                raise RuntimeError("Trellis workspace is missing its FP32 output")
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply sparkinfer/moe/fused_moe/_impl.py:6900]: Already addressed by the existing full-rotation bind initialization. The new capture test now puts bind itself inside torch.cuda.graph, so the caller-owned arena initialization and launch are both replay-validated.

File: tests/moe/test_fused_moe_trellis.py [hunk 8]
```diff
@@ -0,0 +1,847 @@
...
+            w13,
+            w2,
+            gate_suh=gate_suh,
+            up_suh=up_suh,
+            intermediate_rotations=intermediate,
+            down_svh=down_svh,
+            codebook="mul1",
+            tile_config=(64, 128, 64, 128),
+        )
+    with pytest.raises(ValueError, match="unexpected .*MCG marker"):
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/moe/test_fused_moe_trellis.py]: Fixed in 987b829; the regex is now a raw string.

File: tests/moe/test_fused_moe_trellis.py [hunk 9]
```diff
@@ -0,0 +1,847 @@
...
+    """CUDA-graph capture and replay at m strictly below plan capacity.
+
+    This is the surface vLLM #183 actually leaves untested. The existing capture
+    test captures at ``m == max_tokens``, and a small-m eager check does not cover
+    capture: ``bind`` runs inside the captured region, so the route-pack grid, the
+    ``kernel_workspace`` zero-fill and the active-m constants are all baked into
+    the graph. If any of those were captured for the wrong m, replay would produce
+    stale or mismatched results while eager execution stayed correct -- which is
+    exactly a "short prompts pass, long context corrupts" signature once an
+    MTP-N draft replays its captured graph.
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/moe/test_fused_moe_trellis.py:832]: Fixed in 987b829 by moving fused_moe.bind into the torch.cuda.graph block, matching the documented allocation-free/capturable binding contract and testing the bind-time workspace initialization on replay.

**Comment by voipmonitor** [comment]
The requested #49 redesign is already ported here: Trellis now uses the existing fused_moe plan/prepare/scratch/bind/run lifecycle rather than a parallel public API or arena. I preserved your authorship on the implementation commit. Please do not duplicate the port on #49; review and any follow-up should target #90. The latest commit also closes the current CodeRabbit correctness/capture findings and passes the SM120 layer and CUDA-graph tests.

**Comment by voipmonitor** [comment]
Final full-checkpoint validation is complete.

- Pinned checkpoint: `brandonmusic/GLM-5.2-EXL3-TR3-3.0bpw` (295 GiB)
- Configuration: TP4 / DCP4 / MTP3, SM120, CUDA 13.2
- Passed: load, memory profiling, piecewise/full CUDA graph capture, MTP capture, and a real HTTP generation returning `EXL3_OK`
- Capacity at GMU 0.90: 735,232 KV-cache tokens; planned EXL3 prefill arena 414.7 MiB/rank
- SparkInfer regression set: 122 passed

The E2E run found one real planning-contract bug: the scheduler planned `block_m=64`, but execution recomputed 48 from the live `m=1024`, top-k 8, 256-expert shape. That made route packing disagree with its planned expert-block capacity. Commit `2030d39` now carries the planned route geometry through binding and execution, fails early on an explicit launch mismatch, and adds focused regression tests. A direct GPU check confirmed that 48 and 64 produce distinct compiled kernels, so this was not a compile-cache collision.

Paired vLLM consumer: https://github.com/local-inference-lab/vllm/pull/190

**Comment by voipmonitor** [comment]
The core implementation from this PR is now present directly on `master`. Follow-up profiling/capacity fixes are isolated in #92: exact EXL3 arena capacity, complete prewarm, and repeatable KV sizing. #92 is the current SparkInfer review target.
